math/src/main/scala/breeze/linalg/support/CanTraverseValues.scala

```diff
@@ -52,81 +65,105 @@ trait CanTraverseValues[From, A] {
 
 object CanTraverseValues {
 
-  trait ValuesVisitor[@specialized A] {
-    def visit(a: A)
-    def visitArray(arr: Array[A]):Unit = visitArray(arr, 0, arr.length, 1)
+  /**
+    * This trait can be implemented and used to apply a function over a breeze collection/tensor,
+    * using [[CanTraverseValues.traverse()]].
+    *
+    * @tparam V
+    */
+  trait ValuesVisitor[@specialized V] {
 
-    def visitArray(arr: Array[A], offset: Int, length: Int, stride: Int):Unit = {
+    def visit(a: V)
+
+    /** Visits an array assuming offset of zero and stride of 1 */
+    final def visitArray(arr: Array[V]): Unit = visitArray(arr, 0, arr.length, 1)
+
+    def visitArray(arr: Array[V], offset: Int, length: Int, stride: Int): Unit = {
       import spire.syntax.cfor._
       // Standard array bounds check stuff
       if (stride == 1) {
-        cforRange(offset until length + offset) { i =>
-          visit(arr(i))
-        }
+        cforRange(offset until length + offset) { i => visit(arr(i)) }
       } else {
-        cforRange(0 until length) { i =>
-          visit(arr(i * stride + offset))
-        }
+        cforRange(0 until length) { i => visit(arr(i * stride + offset)) }
       }
     }
-    def zeros(numZero: Int, zeroValue: A)
+
+    def zeros(numZero: Int, zeroValue: V): Unit
+
```

Comment:
many operations can be done more quickly on zero values (i.e. you don't
need to visit them all individually), so zeros is a call that Sparse things
can use to tell the ValuesVisitor "simulate the visitation of numZero
zeroValue's"

so the semantics are that it should be equivalent to for(i <- 0 until
numZeros) fn.visit(zeroValue)

On Fri, Apr 8, 2016 at 10:01 AM, Kenta Takagaki <notifications@github.com>
wrote:

> In math/src/main/scala/breeze/linalg/support/CanTraverseValues.scala
> <https://github.com/ktakagaki/breeze/pull/17#discussion_r59056645>:
>
> >        }
> >      }
> > -    def zeros(numZero: Int, zeroValue: A)
> > +
> > +    def zeros(numZero: Int, zeroValue: V): Unit
> > +
>
> @dlwh <https://github.com/dlwh>
> could you give me a hint about exactly what this def zeros() is doing, so
> that I can comment it? It seems like the implementations are somehow
> resetting internal variables of ValuesVisitor implementation classes, as
> in MinVisitor:
>
>             def zeros(numZero: Int, zeroValue: V): Unit = {
>               if(numZero != 0) {
>                 visitedOne = true
>                 min = scala.math.min(zeroValue, min)
>               }
>             }
>
> There are three uses in CSCMatrix, HashVector, and SparseVector, that go
> like:
>
>      /** Iterates all key-value pairs from the given collection. */
>       def traverse(from: CSCMatrix[V], fn: ValuesVisitor[V]): Unit = {
>         fn.zeros(from.size - from.activeSize, from.zero)
>         fn.visitArray(from.data, 0, from.activeSize, 1)
>       }
>     }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ktakagaki/breeze/pull/17/files/0ef0bf906e05475a253f9ef782e4f603688f9063#r59056645>
>